File: k12-edu-microservice/k12-common/src/main/java/com/k12/common/util/CatalogBrowseScope.java
package com.k12.common.util;

import org.springframework.util.StringUtils;

/**
 * 目录节点 → 浏览统计/列表是否包含子树（前后端口径一致）
 */
public final class CatalogBrowseScope {

    private CatalogBrowseScope() {
    }

    /**
     * @param nodeType   目录节点类型
     * @param depth      树深度（0 通常为课本根）
     * @param code       节点编码（如 xxx_root）
     * @param clientHint 前端传入的 includeSubtree，未知 nodeType 时作兜底
     */
    public static boolean resolveIncludeSubtree(
            String nodeType,
            Integer depth,
            String code,
            Boolean clientHint) {
        if (StringUtils.hasText(nodeType)) {
            switch (nodeType.toLowerCase()) {
                case "lesson":
                case "leaf":
                    return false;
                case "unit":
                case "folder":
                case "section":
                case "textbook":
                case "root":
                    return true;
                default:
                    break;
            }
        }
        if (depth != null && depth == 0) {
            return true;
        }
        if (StringUtils.hasText(code) && code.endsWith("_root")) {
            return true;
        }
        return clientHint == null || Boolean.TRUE.equals(clientHint);
    }
}

File: k12-edu-platform/src/utils/catalogBrowseScope.ts
/**
 * 目录节点 → includeSubtree 口径（与后端 CatalogBrowseScope 对齐）
 */
import type { CatalogNode } from '@/types/browse'

/** nodeType → 是否含子树；未列出的 nodeType 走 depth/code/clientHint 兜底 */
export const CATALOG_SUBTREE_BY_NODE_TYPE: Readonly<
  Record<string, boolean | undefined>
> = {
  lesson: false,
  leaf: false,
  unit: true,
  folder: true,
  section: true,
  textbook: true,
  root: true,
}

export function resolveCatalogIncludeSubtree(
  node: Pick<CatalogNode, 'nodeType' | 'depth' | 'code'> | null | undefined,
  clientHint = true,
): boolean {
  if (!node) return clientHint

  const mapped = CATALOG_SUBTREE_BY_NODE_TYPE[node.nodeType]
  if (mapped === false) return false
  if (mapped === true) return true

  if (node.depth === 0) return true
  if (typeof node.code === 'string' && node.code.endsWith('_root')) return true
  return clientHint
}

export function resolveCatalogIncludeSubtreeById(
  findNode: (id: number) => CatalogNode | null | undefined,
  nodeId: number | null | undefined,
  clientHint = true,
): boolean {
  if (nodeId == null || nodeId <= 0) return clientHint
  return resolveCatalogIncludeSubtree(findNode(nodeId), clientHint)
}
